network/src/peer_store/db.rs

```diff
@@ -0,0 +1,313 @@
+use super::{Multiaddr, PeerId, Score, Status};
+use crate::network_group::{Group, NetworkGroup};
+use libp2p::core::Endpoint;
+use rusqlite::types::ToSql;
+use rusqlite::{Connection, NO_PARAMS};
+use std::time::Duration;
+
+#[derive(Debug)]
+pub struct PeerInfo {
+    pub id: u32,
+    pub peer_id: PeerId,
+    pub connected_addr: Multiaddr,
+    pub score: Score,
+    pub status: Status,
+    pub endpoint: Endpoint,
+    pub ban_time: Duration,
+    pub connected_time: Duration,
+}
+
+pub fn create_tables(conn: &Connection) {
+    let sql = r#"
+    CREATE TABLE IF NOT EXISTS peer_info (
+    id INTEGER PRIMARY KEY NOT NULL,
+    peer_id BINARY UNIQUE NOT NULL,
+    connected_addr BINARY NOT NULL,
+    network_group BINARY NOT NULL,
+    score INTEGER NOT NULL,
+    status TEXT NOT NULL,
+    endpoint INTEGER NOT NULL,
+    ban_time INTEGER NOT NULL,
+    connected_time INTEGER NOT NULL
+    );
+    "#;
+    conn.execute_batch(sql).expect("crate peer_info table");
+    let sql = r#"
+    CREATE TABLE IF NOT EXISTS peer_addr (
+    id INTEGER PRIMARY KEY NOT NULL,
+    peer_info_id INTEGER NOT NULL,
+    addr BINARY NOT NULL
+    );
+    CREATE UNIQUE INDEX IF NOT EXISTS idx_peer_info_id_addr_on_peer_addr ON peer_addr (peer_info_id, addr);
+    "#;
+    conn.execute_batch(sql).expect("create peer_addr table");
+    let sql = r#"
+    CREATE TABLE IF NOT EXISTS ban_list (
+    id INTEGER PRIMARY KEY NOT NULL,
+    ip BINARY UNIQUE NOT NULL,
+    ban_time INTEGER NOT NULL
+    );
+    "#;
+    conn.execute_batch(sql).expect("create ban_list table");
+}
+
+pub fn insert_peer_addr(conn: &Connection, peer_info_id: u32, addr: &Multiaddr) -> usize {
+    let mut stmt = conn
+        .prepare(
+            "INSERT OR IGNORE INTO peer_addr (peer_info_id, addr)
+                     VALUES(:peer_info_id, :addr)",
+        )
+        .expect("prepare");
+    stmt.execute_named(&[
+        (":peer_info_id", &peer_info_id),
+        (":addr", &addr.to_bytes()),
+    ])
+    .expect("insert")
+}
+
+pub fn insert_peer_info(
```

Comment:
I don't like code smell here, we need to pass `connection` to each function and many repeated `expect` here. Is it a best practice to use  sqlite lib like this way?

Reply:
Rust can use the trait to extend functions  [[eg]](http://xion.io/post/code/rust-extension-traits.html)

Reply:
Pass connection is rusqlite document recommend style, and the invoker can decide whether connection belongs to a transaction or just a single one. I can't figure a better way to refactor this yet.